src/mam4xx/utils.hpp

```diff
@@ -161,71 +172,145 @@ void transfer_work_arrays_to_prognostics(const Real q[gas_pcnst()],
+
+  if (progs.q_gas[0].data()) { // if gases are defined in dry_aero aerosol state
+    s_idx = gasses_start_ind(); // gases start at index 9 (index 10 in Fortran
+                                // version)
+    for (int g = 0; g < AeroConfig::num_gas_ids(); ++g) {
+      // get mmr at level "klev"
+      q[s_idx] = progs.q_gas[g](klev);
+      s_idx++; // update index
     }
+  } else {
+    s_idx = aero_start_ind(); // If no gasses; start with the first index of
+                              // aerosols
   }
+
+  // Now start adding aerosols mmr into the state_q
+  for (int m = 0; m < AeroConfig::num_modes(); ++m) {
+    // First add the aerosol species mmr
+    for (int a = 0; a < mam4::num_species_mode(m); ++a) {
+      q[s_idx] = progs.q_aero_i[m][a](klev);
+      s_idx++; // update index even if we lack some aerosol mmrs
+    }
+    q[s_idx] = progs.n_mode_i[m](klev);
+    s_idx++; // update index
+  }
+} // state_q_from_progs_at_one_lev
+
+KOKKOS_INLINE_FUNCTION
+void progs_from_state_q_at_one_lev(const Real *q, mam4::Prognostics &progs,
+                                   const int klev) {
+
+  int s_idx = ekat::ScalarTraits<int>::invalid();
+
+  if (progs.q_gas[0].data()) { // if gases are defined in dry_aero aerosol state
+    s_idx = gasses_start_ind(); // gases start at index 9 (index 10 in Fortran
+                                // version)
+    for (int g = 0; g < AeroConfig::num_gas_ids(); ++g) {
+      // get mmr at level "klev"
+      progs.q_gas[g](klev) = q[s_idx];
+      s_idx++; // update index
+    }
+  } else {
+    s_idx = aero_start_ind(); // If no gasses; start with the first index of
+                              // aerosols
+  }                           // end if
+
+  // Now start adding aerosols mmr into the state_q
+  for (int m = 0; m < AeroConfig::num_modes(); ++m) {
+    //   // First add the aerosol species mmr
+    for (int a = 0; a < mam4::num_species_mode(m); ++a) {
+      if (progs.q_aero_i[m][a].data()) {
+        progs.q_aero_i[m][a](klev) = q[s_idx];
+        s_idx++; // update index even if we lack some aerosol mmrs
+      }          // end if
+    }            // a
+    // Now add aerosol number mmr
+    progs.n_mode_i[m](klev) = q[s_idx];
+    s_idx++; // update index
+  }          // m
 }
 
-// converts the quantities in the work arrays vmrs and vmrscw from mass/number
-// mixing ratios to volume/number mixing ratios
+// Given an AerosolState with views for dry aerosol quantities, creates a
+// cloudborne aerosol mmr 1D view for the column with the given index.
+// This object can be provided to mam4xx for the column.
 KOKKOS_INLINE_FUNCTION
-void convert_work_arrays_to_mmr(const Real vmr[gas_pcnst()],
-                                const Real vmrcw[gas_pcnst()],
-                                Real q[gas_pcnst()], Real qqcw[gas_pcnst()]) {
-  DECLARE_PROG_TRANSFER_CONSTANTS
+void qqcw_from_progs_at_one_lev(const mam4::Prognostics &progs, Real *qqcw,
```

Comment:
This could be renamed to something like `extract_qqcw_from_prognostics`, and its inverse could be `inject_qqcw_to_prognostics`, similar to the above suggestions.